GH10559: Minor improvement: Change read_excel sheet name

abarber4gh opened the pull request:
modify io/excel.py and relevant docs (io.rst) to use sheet_name for read_excel
but allow sheetname to still be used for backwards compatibility.  add test_excel
to verify that sheet_name and sheetname args produce the same result.

 - [X] closes #10559 
 - [X] tests added / passed
 - [X] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [x] whatsnew entry

Commit:
GH10559: Minor improvement: Change to_excel sheet name

modify io/excel.py and relevant docs (io.rst) to use sheet_name for read_excel
but allow sheetname to still be used for backwards compatibility.  add test_excel
to verify that sheet_name and sheetname args produce the same result.

jreback commented on pandas/io/excel.py:
```diff
@@ -200,8 +200,12 @@ def read_excel(io, sheetname=0, header=0, skiprows=None, skip_footer=0,
 
+    # maintain backwards compatibility by converting sheetname to sheet_name
+    if 'sheetname' in kwds:
+         sheet_name = kwds.pop('sheetname')
```
needs a deprecation warning on the original arg. You need to *also* accept the original arg as a kwarg.

jreback replied:
you can use ``pandas.util._decorators.deprecate_kwarg``

jreback commented on pandas/tests/io/test_excel.py:
```diff
@@ -544,6 +544,18 @@ def test_date_conversion_overflow(self):
         result = self.get_exceldf('testdateoverflow')
         tm.assert_frame_equal(result, expected)
 
+    # GH10559: Minor improvement: Change to_excel "sheet_name" to "sheetname"
```
comments go inside the function.

jreback commented on pandas/tests/io/test_excel.py:
```diff
@@ -544,6 +544,18 @@ def test_date_conversion_overflow(self):
+    def test_sheet_name_and_sheetname(self):
+        dfref = self.get_csv_refdf('test1')
+        df1 = self.get_exceldf('test1', sheet_name='Sheet1')    # doc
+        df2 = self.get_exceldf('test1', sheetname='Sheet2')     # bkwrds compat
```
this should show a FutureWarning

TomAugspurger replied:
You can assert that a warning with the `tm.assert_produces_warning` context manager.

jreback reviewed (changes requested):
need to add a note in deprecation section

Commit:
GH10559: Minor improvement: Change to_excel sheet name

added @deprecate_kwarg to read_excel as arg changes from sheetname to sheet_name.
moved test comments into function, add assert_produces_warning.

Commit:
GH10559: Minor improvement: Change to_excel sheet name

remove manual arg change, use @deprecate_kwarg to read_excel
as arg changes from sheetname to sheet_name.

Commit:
GH10559: Minor improvement: Change to_excel sheet name

shorten lines under 79 char.

jreback commented on pandas/io/excel.py:
```diff
@@ -48,7 +48,7 @@
     file could be file://localhost/path/to/workbook.xlsx
-sheetname : string, int, mixed list of strings/ints, or None, default 0
+sheet_name : string, int, mixed list of strings/ints, or None, default 0
 
```
can you add ``sheetname (DEPRECATED)`` as well

TomAugspurger replied:
There's also the `deprecated` sphinx directive. I don't see any uses of that, but we can give it a shot here. I think it'd be like

```
sheetname : string, int, mixed list of strings/ints, or None, default 0

    .. deprecated:: 0.21.0
       Use `sheet_name` instead

sheet_name : string, int, mixed list of strings/ints, or None, default 0
```

jreback reviewed (dismissed):
pls add a note in whatsnew (0.21.0) / Deprecations

Commit:
GH10559: Minor improvement: Change to_excel sheet name

update whats new 0.21.0 Deprecations section noting sheetname deprecated
in favor of sheet_name.
add sheetname deprecation in read_excel() docstring.

TomAugspurger commented:
@abarber4gh thanks, nice job! If you want to check the docs http://pandas-docs.github.io/pandas-docs-travis/generated/pandas.read_excel.html when [this build](https://travis-ci.org/pandas-dev/pandas/jobs/235365169) finishes, I'm curious to see how the `.. deprecated:` directive looks.
